File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/camera/AnimationOptions.java
package org.maplibre.nativeffi.camera;

import java.util.Objects;

/** Mutable descriptor for animated camera transitions. */
public final class AnimationOptions {
  private Double durationMs;
  private Double velocity;
  private Double minZoom;
  private UnitBezier easing;

  public boolean hasDurationMs() {
    return durationMs != null;
  }

  public Double durationMs() {
    return durationMs;
  }

  public AnimationOptions durationMs(double durationMs) {
    this.durationMs = durationMs;
    return this;
  }

  public AnimationOptions clearDurationMs() {
    durationMs = null;
    return this;
  }

  public boolean hasVelocity() {
    return velocity != null;
  }

  public Double velocity() {
    return velocity;
  }

  public AnimationOptions velocity(double velocity) {
    this.velocity = velocity;
    return this;
  }

  public AnimationOptions clearVelocity() {
    velocity = null;
    return this;
  }

  public boolean hasMinZoom() {
    return minZoom != null;
  }

  public Double minZoom() {
    return minZoom;
  }

  public AnimationOptions minZoom(double minZoom) {
    this.minZoom = minZoom;
    return this;
  }

  public AnimationOptions clearMinZoom() {
    minZoom = null;
    return this;
  }

  public boolean hasEasing() {
    return easing != null;
  }

  public UnitBezier easing() {
    return easing;
  }

  public AnimationOptions easing(UnitBezier easing) {
    this.easing = Objects.requireNonNull(easing, "easing");
    return this;
  }

  public AnimationOptions clearEasing() {
    easing = null;
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/camera/AnimationOptions.kt
package org.maplibre.nativeffi.camera

/**
 * Mutable animation descriptor used for animated camera commands.
 *
 * Compares and hashes by field value; [copy] returns an independent instance. Keep an instance
 * unmodified while it is a key in a hash-based collection.
 */
public class AnimationOptions {
  public var durationMs: Double? = null

  public var velocity: Double? = null

  public var minZoom: Double? = null

  public var easing: UnitBezier? = null

  /**
   * Caller-chosen identity for the transition these options start. Leaving it null leaves the
   * transition silent.
   *
   * When set, the transition emits one
   * [org.maplibre.nativeffi.runtime.RuntimeEventType.MAP_CAMERA_TRANSITION_FINISHED] event carrying
   * this value, however the transition ends: completed, superseded, cancelled, or applied instantly
   * as a zero-duration jump. The event reports only that the transition released the camera, not
   * the outcome. A rejected command starts no transition and emits no event.
   *
   * The value is opaque to MapLibre Native. Native `uint64_t` carried as a [Long] bit pattern.
   */
  public var transitionId: Long? = null

  /** Returns an independent copy of this descriptor with [block] applied to the copy. */
  public fun copy(block: AnimationOptions.() -> Unit = {}): AnimationOptions =
    AnimationOptions()
      .also {
        it.durationMs = durationMs
        it.velocity = velocity
        it.minZoom = minZoom
        it.easing = easing
        it.transitionId = transitionId
      }
      .apply(block)

  private val fields: List<Any?>
    get() = listOf(durationMs, velocity, minZoom, easing, transitionId)

  override fun equals(other: Any?): Boolean = other is AnimationOptions && fields == other.fields

  override fun hashCode(): Int = fields.hashCode()
}
